### PR TITLE
Unit tests miscellaneous patches

### PR DESCRIPTION
**Pull request by MeatBoy106**
Compiling and running unit tests with clang-16 and address sanitizer exhibits errors related to old-style memory management or permissive syntax. This pull request addresses these issues and provides opportunistic improvements along the way.

### CODE REVIEW

**Comment by MeatBoy106** [comment]
I suggest enabling address sanitizer for unit tests as it helps catching subtle false negative, memory leaks and so. I'm not sure how to provide such patch myself though.

File: src/base/inifile.cc [hunk 1]
```diff
@@ -294,32 +285,29 @@ IniFile::Section::printUnreferenced(const std::string &sectionName)
...
+IniFile::printUnreferenced() const
 {
     bool unref = false;
 
-    for (SectionTable::iterator i = table.begin();
-         i != table.end(); ++i) {
-        const std::string &sectionName = i->first;
-        Section *section = i->second;
+    for (auto& entry: table) {
+        auto& [sectionName, section] = entry;
```
Comment (powerjg) [inline src/base/inifile.cc:300]: Do we know when GCC/clang started supporting this syntax? I want to make sure that all of our supported compilers support this.

**Review by powerjg — COMMENTED** [review]
Generally, this looks good to me. There are two things that we might want to do before merging.
1. It would be great to run the compiler tests on this before merging... I think this might be possible in github now.
2. For the trie... this was implemented to have *very* high performance lookups. We should make sure that this change doesn't have an impact on gem5's performance. (On a side note, I personally and not convinced that the trie is really needed for performance, so this may be a moot issue.)

**Review by BobbyRBruce — COMMENTED** [review]
Please address Jason's question

(Starting a review here so this comes up as being "under review" rather than "pending review")

File: src/base/inifile.cc (hunk 1 above)
Comment (MeatBoy106) [reply src/base/inifile.cc:300]: image

**Comment by MeatBoy106** [comment]
Regarding performance, this new implementation guarantees the same number of memory allocation and pointer indirections than the previous implementation. The only theoretical difference I can think of should come from the unique_ptr move that implies reseting the moved pointer after it has been moved from. I would not be surprised if optimization removed both the reset and the destructor call for move unique_ptr though. My guess is: no measurable difference.

**Review by powerjg — APPROVED** [review]
Thanks for looking into the compilers and the other question :)

**Pull request merged**